"You are trying to create client-side redirections to paths that do not exist" to folders without trailing slash

### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [ ] I have tried the `npm run clear` or `yarn clear` command.
- [ ] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

For paths to folders without a trailing slash, hosts will typically 302 redirect to the path with the slash

e.g. https://docs.sheetjs.com/docs/getting-started redirects to https://docs.sheetjs.com/docs/getting-started/

`@docusaurus/plugin-client-redirects` does not seem to be aware of this behavior for folders that have subfolders.

This redirect works

```js
        { from: '/docs/installation', to: '/docs/getting-started/' }, // works
```

This redirect elicits the error mentioned in the title:

```js
        { from: '/docs/installation', to: '/docs/getting-started' }, // elicits error
```

The fact that it works for leaf folders that only have `index.html` suggests this is a bug in the redirect scanner:

```js
        { from: '/docs/example', to: '/docs/getting-started/example' }, // this works, even though `example` is a folder
```



### Reproducible demo

https://github.com/SheetJS/docs.sheetjs.com

### Steps to reproduce

```bash
git clone --depth=1 https://github.com/SheetJS/docs.sheetjs.com
cd docs.sheetjs.com/docz
npm i
npm run build ## note: this should work
sed -I '' "s#started/'#started'#g" docusaurus.config.js
npm run build ## this will fail with the aforementioned error
```

If the `sed` command fails, the edit is in `docusaurus.config.js`:

```diff
       redirects: [
         { from: '/docs/example', to: '/docs/getting-started/example' },
-        { from: '/docs/installation', to: '/docs/getting-started/' },
+        { from: '/docs/installation', to: '/docs/getting-started' },
       ]
     }]
```

### Expected behavior

It should create the redirect

### Actual behavior

```
[ERROR] Error: You are trying to create client-side redirections to paths that do not exist:
```

The path in question is not listed in the valid paths

### Your environment

_No response_

### Self-service

- [X] I'd be willing to fix this bug myself.

> For paths to folders without a trailing slash, hosts will typically 302 redirect to the path with the slash

That's not 100% correct. GitHub Pages does it, not all others, or at least it depends on their configuration.

That would be worth mentioning that you are using GitHub Pages + Cloudflare on top (seen that in request headers).

Note our doc recommends to use an explicit `trailingSlash` setting for GitHub Pages:

https://docusaurus.io/docs/deployment#deploying-to-github-pages

<img width="716" alt="CleanShot 2022-08-31 at 16 27 05@2x" src="https://user-images.githubusercontent.com/749374/187703254-e0a3bd92-77f4-4222-8024-3b3a477f4ef7.png">

In my experience, this is particularly important if you use relative links such as `[xyz](./xyz)` (which we also don't particularly advice) where the link would resolve differently according to the presence/absence of a trailing slash in the current URL.

And you'll notice that you can in practice browse the page with both URLs:
- GitHub pages add a trailing slash on the initial page load: `https://docs.sheetjs.com/docs/getting-started/example/`
- Docusaurus will not add a trailing slash after any SPA navigation: `https://docs.sheetjs.com/docs/getting-started/example`

Really, you'd better to use an explicit setting.

---

> [ERROR] Error: You are trying to create client-side redirections to paths that do not exist:

That would be convenient to paste the full error message instead of just the beginning 😅 

---

> One way to resolve the issue without disrupting other plugins is to add both slash variants to the validation.

> I think for trailingSlash: undefined only, we should accept both alternatives.

👍 seems reasonable

--- a/packages/docusaurus-plugin-client-redirects/src/__tests__/collectRedirects.test.ts
+++ b/packages/docusaurus-plugin-client-redirects/src/__tests__/collectRedirects.test.ts
@@ -220,6 +220,69 @@ describe('collectRedirects', () => {
     ).toThrowErrorMatchingSnapshot();
   });
 
+  it('tolerates mismatched trailing slash if option is undefined', () => {
+    expect(
+      collectRedirects(
+        createTestPluginContext(
+          {
+            redirects: [
+              {
+                from: '/someLegacyPath',
+                to: '/somePath',
+              },
+            ],
+          },
+          ['/', '/somePath/'],
+          {trailingSlash: undefined},
+        ),
+        undefined,
+      ),
+    ).toEqual([
+      {
+        from: '/someLegacyPath',
+        to: '/somePath',
+      },
+    ]);
+  });
+
+  it('throw if plugin option redirects contain to paths with mismatching trailing slash', () => {
+    expect(() =>
+      collectRedirects(
+        createTestPluginContext(
+          {
+            redirects: [
+              {
+                from: '/someLegacyPath',
+                to: '/someExistingPath/',
+              },
+            ],
+          },
+          ['/', '/someExistingPath', '/anotherExistingPath'],
+          {trailingSlash: false},
+        ),
+        undefined,
+      ),
+    ).toThrowErrorMatchingSnapshot();
+
+    expect(() =>
+      collectRedirects(
+        createTestPluginContext(
+          {
+            redirects: [
+              {
+                from: '/someLegacyPath',
+                to: '/someExistingPath',
+              },
+            ],
+          },
+          ['/', '/someExistingPath/', '/anotherExistingPath/'],
+          {trailingSlash: true},
+        ),
+        undefined,
+      ),
+    ).toThrowErrorMatchingSnapshot();
+  });
+
   it('collects redirects with custom redirect creator', () => {
     expect(
       collectRedirects(

--- a/packages/docusaurus-plugin-client-redirects/src/collectRedirects.ts
+++ b/packages/docusaurus-plugin-client-redirects/src/collectRedirects.ts
@@ -7,6 +7,7 @@
 
 import _ from 'lodash';
 import logger from '@docusaurus/logger';
+import {addTrailingSlash, removeTrailingSlash} from '@docusaurus/utils';
 import {applyTrailingSlash} from '@docusaurus/utils-common';
 import {
   createFromExtensionsRedirects,
@@ -80,16 +81,59 @@ function validateCollectedRedirects(
 
   const allowedToPaths = pluginContext.relativeRoutesPaths;
   const toPaths = redirects.map((redirect) => redirect.to);
-  const illegalToPaths = _.difference(toPaths, allowedToPaths);
-  if (illegalToPaths.length > 0) {
-    throw new Error(
-      `You are trying to create client-side redirections to paths that do not exist:
-- ${illegalToPaths.join('\n- ')}
+  const trailingSlashConfig = pluginContext.siteConfig.trailingSlash;
+  // Key is the path, value is whether a valid toPath with a different trailing
+  // slash exists; if the key doesn't exist it means it's valid
+  const differByTrailSlash = new Map(toPaths.map((path) => [path, false]));
+  allowedToPaths.forEach((toPath) => {
+    if (differByTrailSlash.has(toPath)) {
+      differByTrailSlash.delete(toPath);
+    } else if (differByTrailSlash.has(removeTrailingSlash(toPath))) {
+      if (trailingSlashConfig === true) {
+        differByTrailSlash.set(removeTrailingSlash(toPath), true);
+      } else {
+        differByTrailSlash.delete(removeTrailingSlash(toPath));
+      }
+    } else if (differByTrailSlash.has(addTrailingSlash(toPath))) {
+      if (trailingSlashConfig === false) {
+        differByTrailSlash.set(addTrailingSlash(toPath), true);
+      } else {
+        differByTrailSlash.delete(addTrailingSlash(toPath));
+      }
+    }
+  });
+  if (differByTrailSlash.size > 0) {
+    console.log(differByTrailSlash);
+    const errors = Array.from(differByTrailSlash.entries());
+
+    let message =
+      'You are trying to create client-side redirections to invalid paths.\n';
+
+    const [trailingSlashIssues, invalidPaths] = _.partition(
+      errors,
+      ([, differ]) => differ,
+    );
+
+    if (trailingSlashIssues.length) {
+      message += `
+These paths do exist, but because you have explicitly set trailingSlash=${trailingSlashConfig}, you need to write the path ${
+        trailingSlashConfig ? 'with trailing slash' : 'without trailing slash'
+      }:
+- ${trailingSlashIssues.map(([p]) => p).join('\n- ')}
+`;
+    }
+
+    if (invalidPaths.length) {
+      message += `
+These paths are redirected to but do not exist:
+- ${invalidPaths.map(([p]) => p).join('\n- ')}
 
 Valid paths you can redirect to:
 - ${allowedToPaths.join('\n- ')}
-`,
-    );
+`;
+    }
+
+    throw new Error(message);
   }
 }
 

